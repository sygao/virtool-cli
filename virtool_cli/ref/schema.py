from pydantic import BaseModel, computed_field

from virtool_cli.utils.models import Molecule


class OTUSchema(BaseModel):
    """A schema for the intended data"""

    molecule: Molecule
    """The molecular metadata for this OTU."""

    segments: dict[str, bool]
    """The segments contained in this OTU."""

    @computed_field
    def multipartite(self) -> bool:
        if len(self.segments) < 2:
            return False

        return True
