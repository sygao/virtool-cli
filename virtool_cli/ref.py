import sys
from pathlib import Path
import click
import structlog

import virtool_cli.build
import virtool_cli.divide
import virtool_cli.isolate
import virtool_cli.repair
import virtool_cli.taxid
from virtool_cli.migrate import run as run_migrate
from virtool_cli.update import run as run_update
from virtool_cli.accessions.catalog import run as run_catalog

ERROR_MESSAGE = click.style("ERROR: ", fg='red')

logger = structlog.get_logger()
    
@click.group("ref")
def ref():
    """
    Commands related to reference files.
    """
    pass


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a database reference directory",
)
@click.option(
    "-o",
    "--output",
    type=str,
    default="reference.json",
    help="the output path for a reference.json file",
)
@click.option("-i", "--indent", is_flag=True)
@click.option(
    "-V",
    "--version",
    default=None,
    type=str,
    help="the version string to include in the reference.json file",
)
def build(src_path, output, indent, version):
    """Build a Virtool reference JSON file from a source directory."""
    build_logger = logger.bind(command='build', src_path=src_path, out_path=output)

    if not Path(src_path).exists():
        build_logger.error('Directory not found at src_path')
        return
    
    try:
        virtool_cli.build.run(Path(src_path), Path(output), indent, version)
    except (FileNotFoundError, NotADirectoryError):
        build_logger.exception('Source directory has critical errors')
    except:
        build_logger.exception('Unexpected exception')


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a input reference.json file",
)
@click.option(
    "-o",
    "--output",
    default="src",
    type=str,
    help="the output path for a divided reference directory tree",
)
def divide(src_path, output):
    """Divide a reference.json file from Virtool into a reference directory tree."""
    try:
        if not src_path.endswith(".json"):
            raise TypeError
        virtool_cli.divide.run(Path(src_path), Path(output))
    except (TypeError, FileNotFoundError) as e:
        click.echo(
            ERROR_MESSAGE + "Specified reference file either does not exist or is not a proper JSON file"
        )
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to the input reference.json file",
)
@click.option("-f", "--force_update", is_flag=True)
def taxid(src_path, force_update):
    """Fetch taxid for all OTU in given reference directory."""
    try:
        click.echo("Scanning OTUs for taxon IDs...")
        virtool_cli.taxid.run(Path(src_path), force_update)
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def isolate(src_path):
    """Fetch new isolates for all OTU in a given reference directory."""
    try:
        virtool_cli.isolate.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo("Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def repair(src_path):
    """Fix every OTU in a given reference directory."""
    if not Path(src_path).exists():
        logger.critical('Source directory does not exist')

    try:
        virtool_cli.repair.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option('--debug/--no-debug', default=False)
def migrate(src_path, debug):
    """Convert a reference directory from v1.x to v2.x"""
    if not Path(src_path).exists():
        logger.critical('Source directory does not exist')

    try:
        run_migrate(Path(src_path))
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        click.echo(e)


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option(
    "-cat",
    "--catalog_path",
    required=True,
    type=str,
    default='.cache/catalog',
    help="the path to a catalog directory",
)
@click.option('--debug/--no-debug', default=False)
def catalog(src_path, catalog_path, debug):
    """Update or generate a catalog of all included accessions in a src directory"""
    if not Path(src_path).exists():
        click.echo(ERROR_MESSAGE + 'Source directory does not exist')
        return
    
    catalog_dir = Path(catalog_path)
    if not catalog_dir.exists():
        catalog_dir.mkdir(parents=True)

    try:
        run_catalog(
            src_path=Path(src_path),
            catalog_path=Path(catalog_path),
            debugging=debug
        )
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo(ERROR_MESSAGE + "Not a valid reference directory")
        logger.exception(e)


@ref.command()
@click.option(
    "-src", "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
@click.option(
    "-cat",
    "--catalog_path",
    required=True,
    type=str,
    default='.cache/catalog',
    help="the path to a catalog directory",
)
@click.option('--debug/--no-debug', default=False)
def update(src_path, catalog_path, debug):
    """Fetch new isolates for all OTU in a given reference directory."""
    if not Path(catalog_path).exists():
        click.echo("Not a valid catalog directory")
        return

    try:
        run_update(Path(src_path), Path(catalog_path), debugging=debug)
    except (FileNotFoundError, NotADirectoryError) as e:
        click.echo("Not a valid reference directory")
        click.echo(e)

if __name__ == "__main__":
    ref()
