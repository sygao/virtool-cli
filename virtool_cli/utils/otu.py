import json
from pathlib import Path

import aiofiles

from virtool_cli.utils.reference import get_isolate_paths, get_sequence_paths


async def get_otu_accessions(otu_path: Path) -> list:
    """
    Gets all accessions from an OTU directory and returns a list

    :param otu_path: Path to an OTU directory under a reference directory
    :return: A list of all accessions under an OTU
    """
    accession_list = []

    for isolate_path in get_isolate_paths(otu_path):
        for sequence_path in get_sequence_paths(isolate_path):
            with open(sequence_path, "r") as f:
                sequence = json.load(f)

            accession_list.append(
                sequence.get('accession')
            )

    return accession_list


async def label_isolates(otu_path: Path) -> dict:
    """
    Return all isolates present in an OTU directory

    :param otu_path: Path to an OTU directory
    :return: A dictionary of isolates indexed by source_name
    """
    if not otu_path.exists():
        raise FileNotFoundError

    isolates = {}

    for iso_path in get_isolate_paths(otu_path):
        async with aiofiles.open(iso_path / "isolate.json", "r") as f:
            contents = await f.read()
            isolate = json.loads(contents)
        isolates[isolate.get("source_name")] = isolate

    return isolates


async def get_otu_accessions_metadata(otu_path: Path) -> dict:
    """
    Returns sequence metadata for all sequences present under an OTU

    :param otu_path: Path to an OTU directory under a reference directory
    :return: An accession-keyed dict containing all constituent sequence metadata
    """
    # get length and segment metadata from sequences
    all_metadata = {}

    for isolate_path in get_isolate_paths(otu_path):
        for sequence_path in get_sequence_paths(isolate_path):
            sequence_metadata = await get_sequence_metadata(sequence_path)

            accession = sequence_metadata["accession"]

            all_metadata[accession] = sequence_metadata

    return all_metadata


async def read_otu(path: Path) -> dict:
    """
    Returns a json file in dict form

    :param path: Path to an OTU directory under a reference source
    :return: Deserialized OTU data in dict form
    """
    async with aiofiles.open(path / "otu.json", "r") as f:
        contents = await f.read()
        otu = json.loads(contents)

    return otu


async def fetch_exclusions(otu_path: Path) -> list:
    async with aiofiles.open(otu_path / "exclusions.json", "r") as f:
        contents = await f.read()
        exclusions = json.loads(contents)

    return exclusions


async def get_sequence_metadata(sequence_path: Path) -> dict:
    """
    Gets the accession length and segment name from a sequence file
    and returns it in a dict

    :param sequence_path: Path to a sequence file
    :return: A dict containing the sequence accession, sequence length and segment name if present
    """
    sequence = await parse_sequence(sequence_path)

    sequence_metadata = {
        "accession": sequence["accession"],
        "length": len(sequence["sequence"]),
    }

    segment = sequence.get("segment", None)
    if segment is not None:
        sequence_metadata["segment"] = segment

    return sequence_metadata


async def parse_sequence(path):
    async with aiofiles.open(path, "r") as f:
        contents = await f.read()
        sequence = json.loads(contents)

    return sequence
