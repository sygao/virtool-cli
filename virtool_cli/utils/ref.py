import json
import re
from pathlib import Path


def get_otu_paths(src_path: Path) -> list:
    """
    Generates a list of paths to all OTUs in a src directory.

    :param src_path: Path to a src database directory
    :return: List of paths to all OTU in a src directory
    """
    return [otu for otu in src_path.iterdir() if otu.is_dir()]

def generate_otu_dirname(otu_name: str):
    """
    """
    no_whitespace = otu_name.replace(" ", "_")
    no_plus = no_whitespace.replace('+', 'and')
    no_symbols = re.split(r'[()/-]+', no_plus)

    return no_symbols

def parse_otu(path: Path) -> dict:
    """
    Returns a json file in dict form

    :param paths: Path to an OTU in a reference
    :return: OTU data in dict form
    """
    with open(path / "otu.json", "r") as f:
        otu = json.load(f)

    return otu