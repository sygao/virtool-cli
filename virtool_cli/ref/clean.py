import shutil
from pathlib import Path
import structlog
import click

from virtool_cli.utils.logging import configure_logger

base_logger = structlog.get_logger()


def run(repo_path: Path, debugging: bool = False):
    click.echo("Cleaning cache files...")
    cache_path = repo_path / ".cache"
    clean_cache(cache_path)


def clean_cache(cache_path):
    if cache_path.exists():
        shutil.rmtree(cache_path / 'updates')
