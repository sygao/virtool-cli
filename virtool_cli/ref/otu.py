import sys
from collections import defaultdict

import structlog

from virtool_cli.ncbi.client import NCBIClient
from virtool_cli.ncbi.model import NCBIGenbank
from virtool_cli.ref.repo import EventSourcedRepo
from virtool_cli.ref.resources import EventSourcedRepoOTU
from virtool_cli.ref.schema import OTUSchema
from virtool_cli.ref.utils import IsolateName, IsolateNameType
from virtool_cli.utils.models import Molecule

logger = structlog.get_logger("otu")


def create_otu_auto(
    repo: EventSourcedRepo,
    taxid: int,
    accessions: list[str],
    ignore_cache: bool = False,
):
    otu_logger = logger.bind(taxid=taxid)

    client = NCBIClient.from_repo(repo.path, ignore_cache)

    taxonomy = client.fetch_taxonomy_record(taxid)
    if taxonomy is None:
        otu_logger.fatal(f"Could not retrieve {taxid} from NCBI Taxonomy")
        return None

    records = client.fetch_genbank_records(accessions)
    if len(records) != len(accessions):
        otu_logger.fatal(f"Could not retrieve all requested accessions.")
        return None

    binned_records = group_genbank_records_by_isolate(records)
    if len(binned_records) > 1:
        otu_logger.fatal(
            f"More than one isolate found. Cannot create schema automatically."
        )
        return None


def create_otu(
    repo: EventSourcedRepo,
    taxid: int,
    ignore_cache: bool = False,
) -> EventSourcedRepoOTU:
    """Initialize a new OTU from a Taxonomy ID."""
    otu_logger = logger.bind(taxid=taxid)

    if repo.get_otu_by_taxid(taxid):
        raise ValueError(
            f"Taxonomy ID {taxid} has already been added to this reference.",
        )

    ncbi = NCBIClient.from_repo(repo.path, ignore_cache)

    taxonomy = ncbi.fetch_taxonomy_record(taxid)

    if taxonomy is None:
        otu_logger.fatal(f"Taxonomy ID {taxid} not found")
        sys.exit(1)

    try:
        otu = repo.create_otu(
            acronym="",
            legacy_id=None,
            name=taxonomy.name,
            schema=None,
            taxid=taxid,
        )

        return otu
    except ValueError as e:
        otu_logger.warning(e)
        sys.exit(1)


def update_otu(
    repo: EventSourcedRepo,
    otu: EventSourcedRepoOTU,
    ignore_cache: bool = False,
):
    """Fetch a full list of Nucleotide accessions associated with the OTU
    and pass the list to the add method.
    """
    ncbi = NCBIClient.from_repo(repo.path, ignore_cache)

    linked_accessions = ncbi.link_accessions_from_taxid(otu.taxid)

    add_sequences(repo, otu, linked_accessions)


def add_sequences(
    repo: EventSourcedRepo,
    otu: EventSourcedRepoOTU,
    accessions: list,
    ignore_cache: bool = False,
):
    """Take a list of accessions, filter for eligible accessions and
    add new sequences to the OTU
    """
    client = NCBIClient.from_repo(repo.path, ignore_cache)

    otu_logger = logger.bind(taxid=otu.taxid, otu_id=str(otu.id), name=otu.name)
    fetch_list = list(set(accessions).difference(otu.blocked_accessions))
    if not fetch_list:
        otu_logger.info("OTU is up to date.")
        return

    otu_logger.info(f"Fetching {len(fetch_list)} accessions...", fetch_list=fetch_list)

    records = client.fetch_genbank_records(fetch_list)

    # if records and not otu.molecule:
    #     # TODO: Upcoming UpdateMolecule event?
    #     molecule = get_molecule_from_records(records)
    #     otu_logger.debug("Retrieved new molecule data", molecule=molecule)

    record_bins = group_genbank_records_by_isolate(records)

    new_accessions = []

    for isolate_key in record_bins:
        record_bin = record_bins[isolate_key]

        isolate_id = otu.get_isolate_id_by_name(isolate_key)
        if isolate_id is None:
            otu_logger.debug(
                f"Creating isolate for {isolate_key.type}, {isolate_key.value}",
            )
            isolate = repo.create_isolate(
                otu_id=otu.id,
                legacy_id=None,
                source_name=isolate_key.value,
                source_type=isolate_key.type,
            )
            isolate_id = isolate.id

        for accession in record_bin:
            record = record_bin[accession]
            if accession in otu.accessions:
                otu_logger.warning(f"{accession} already exists in OTU")
            else:
                sequence = repo.create_sequence(
                    otu_id=otu.id,
                    isolate_id=isolate_id,
                    accession=record.accession,
                    definition=record.definition,
                    legacy_id=None,
                    segment=record.source.segment,
                    sequence=record.sequence,
                )

                new_accessions.append(sequence.accession)

    if new_accessions:
        otu_logger.info(
            f"Added {len(new_accessions)} sequences to {otu.taxid}",
            new_accessions=new_accessions,
        )

    else:
        otu_logger.info("No new sequences added to OTU")


def create_schema_from_records(records: list[NCBIGenbank]) -> OTUSchema | None:
    molecule = get_molecule_from_records(records)

    binned_records = group_genbank_records_by_isolate(records)
    if len(binned_records) > 1:
        logger.fatal(
            "More than one isolate found. Cannot create schema automatically.",
            bins=binned_records,
        )
        return None

    if segments is None:
        segments = _get_segments_from_records(records)

    if segments:
        return OTUSchema(molecule=molecule, segments=segments)

    return None


def get_molecule_from_records(records: list[NCBIGenbank]) -> Molecule:
    """Return relevant molecule metadata from one or more records"""
    if not records:
        raise IndexError("No records given")

    for record in records:
        if record.refseq:
            return Molecule.model_validate(
                {
                    "strandedness": record.strandedness.value,
                    "type": record.moltype.value,
                    "topology": record.topology.value,
                }
            )

    record = records[0]
    return Molecule.model_validate(
        {
            "strandedness": record.strandedness.value,
            "type": record.moltype.value,
            "topology": record.topology.value,
        }
    )


def group_genbank_records_by_isolate(
    records: list[NCBIGenbank],
) -> dict[IsolateName, dict[str, NCBIGenbank]]:
    """Indexes Genbank records by isolate name"""
    isolates = defaultdict(dict)

    for record in records:
        if (isolate_name := _get_isolate_name(record)) is not None:
            isolates[isolate_name][record.accession] = record

    return isolates


def _get_isolate_name(record: NCBIGenbank) -> IsolateName | None:
    """Get the isolate name from a Genbank record"""
    record_logger = logger.bind(
        accession=record.accession,
        definition=record.definition,
        source_data=record.source,
    )

    if record.source.model_fields_set.intersection(
        {IsolateNameType.ISOLATE, IsolateNameType.STRAIN, IsolateNameType.CLONE},
    ):
        for source_type in IsolateNameType:
            if source_type in record.source.model_fields_set:
                return IsolateName(
                    type=IsolateNameType(source_type),
                    value=record.source.model_dump()[source_type],
                )

    elif record.refseq:
        record_logger.debug(
            "RefSeq record does not contain sufficient source data. "
            + "Edit before inclusion."
        )

        return IsolateName(
            type=IsolateNameType(IsolateNameType.REFSEQ),
            value=record.accession,
        )

    record_logger.debug("Record does not contain sufficient source data for inclusion.")

    return None


def _get_segments_from_records(records) -> dict[str, bool]:
    if len(records) == 1:
        record = records[0]

        if record.source.segment != "":
            segment_name = record.source.segment
        else:
            segment_name = record.source.organism

        return {segment_name: True}

    segment_set = set()
    for record in records:
        if record.source.segment:
            segment_set.add(record.source.segment)
        else:
            raise ValueError("No segment name found for multipartite OTU segment.")

    return {segment_name: True for segment_name in segment_set}
