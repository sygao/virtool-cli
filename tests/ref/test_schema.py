import pytest
from syrupy import SnapshotAssertion

from virtool_cli.ref.schema import OTUSchema
from virtool_cli.ref.otu import create_schema_from_records


@pytest.mark.parametrize(
    "accessions",
    [
        ["NC_024301"],
        [
            "NC_010314",
            "NC_010315",
            "NC_010316",
            "NC_010317",
            "NC_010318",
            "NC_010319",
        ],
    ],
)
def test_create_schema_from_records(
    accessions: list[str], scratch_ncbi_client, snapshot: SnapshotAssertion
):
    records = scratch_ncbi_client.fetch_genbank_records(accessions)
    auto_schema = create_schema_from_records(records)

    assert type(auto_schema) is OTUSchema

    assert auto_schema.model_dump() == snapshot
