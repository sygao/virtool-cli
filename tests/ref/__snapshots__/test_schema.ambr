# serializer version: 1
# name: test_create_schema_from_records[accessions0]
  dict({
    'molecule': dict({
      'strandedness': <Strandedness.DOUBLE: 'double'>,
      'topology': <Topology.CIRCULAR: 'circular'>,
      'type': <MolType.DNA: 'DNA'>,
    }),
    'multipartite': False,
    'segments': dict({
      'Pagoda yellow mosaic associated virus': True,
    }),
  })
# ---
# name: test_create_schema_from_records[accessions1]
  dict({
    'molecule': dict({
      'strandedness': <Strandedness.SINGLE: 'single'>,
      'topology': <Topology.CIRCULAR: 'circular'>,
      'type': <MolType.DNA: 'DNA'>,
    }),
    'multipartite': True,
    'segments': dict({
      'DNA C': True,
      'DNA M': True,
      'DNA N': True,
      'DNA R': True,
      'DNA S': True,
      'DNA U3': True,
    }),
  })
# ---
